# 0014-variables-array.patch
#
# DE DONDE SALE. Parche 0008 de los 37 de syntax-transform/3.1/ del PR https://github.com/mkxp-z/mkxp-z/pull/304
# de mkxp-z, de white-axe (rama syntax-transform, 0be32ad119b054749cf3db8c72434f8fa1ac83ea).
# Ensenan al parser y al nucleo de Ruby 3.1 la gramatica y la semantica de
# Ruby 1.8, que es en lo que estan escritos cinco de los ocho juegos del censo.
#
# COMO SE APLICA. Sobre src/ruby-3.1.7 DESPUES de los parches 0001-0005 y en
# orden de numero: el 0006 (prelude) monta el mecanismo que los otros 36 usan,
# y varios de ellos tocan el mismo fichero, asi que el orden no es decorativo.
#
#     cd src/ruby-3.1.7
#     sed -i 's/\r$//' string.c
#     for n in $(seq 6 42); do
#       patch -p1 -i "../../patches/$(printf '%04d' $n)-"*.patch || break
#     done
#     rm -f parse.c id.h insns.inc insns_info.inc opt_sc.inc optinsn.inc \
#           optunifs.inc vm.inc vmtc.inc mjit_compile.inc
#
# Ese `rm` NO es opcional y cuesta una tarde descubrirlo. El tarball de Ruby
# trae parse.c y los .inc YA GENERADOS en el propio directorio de fuentes, y
# un `#include "insns.inc"` desde compile.c mira primero la carpeta del fichero
# que incluye: aunque el build regenere los suyos al lado del Makefile, gana el
# viejo. Con parse.c viejo los ocho parches de gramatica no llegan al binario y
# NADA lo dice; con los .inc viejos, al menos, no compila
# ('YARVINSN_mkxp_syntax_transform_raise_invalid_break' undeclared).
#
# El cuerpo del diff, de aqui abajo, es el del PR VERBATIM: los 37 aplicaron
# sin fuzz y sin rechazos sobre este arbol, asi que no habia nada que adaptar.
# Lo unico que hay que vigilar es string.c, que en este repositorio esta en
# CRLF y al que `patch` le deja el fichero entero en LF; se normaliza antes,
# para que ese cambio sea deliberado y no un efecto secundario.
#
# COMPROBADO. Con los 37 puestos, bison 3.8.2 acepta la gramatica (parse.y
# declara %expect 0: cualquier conflicto nuevo habria sido un error, y no hubo
# ninguno) y CRuby 3.1.7 cruza-compila para aarch64 con devkitA64 16.1.0 hasta
# libruby-static.a (44,7 MB), con los simbolos del mecanismo dentro
# (mkxp_syntax_transform_set_next_eval, mkxp_ec_is_syntax_transform_active,
# mkxp_syntax_transform_target_ruby_version_*) y 14 instrucciones nuevas en
# insns.inc. No se ha ejecutado: eso pide consola.
#
# QUE HACE FALTA ADEMAS. Estos parches por si solos NO cambian el
# comportamiento: mkxp_syntax_transform_target_ruby_version_* nace en -1 y la
# transformacion queda apagada, asi que el interprete se comporta como el de
# hoy. Encenderla es cosa del lado de mkxp-z (armar
# mkxp_syntax_transform_set_next_eval(1) alrededor del eval de cada seccion de
# Scripts.rxdata y fijar la version objetivo desde mkxp.json), que en este
# repositorio todavia no esta hecho.
#
# Licencia: GPL-2.0-or-later, la del PR. La cabecera original va debajo.
# ---------------------------------------------------------------------------

#
# This file is part of mkxp.
#
# Copyright (C) 2026 The mkxp-z authors
#
# mkxp is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# mkxp is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with mkxp.  If not, see <http://www.gnu.org/licenses/>.

# In Ruby 1.8, all methods that return an array of the names of variables or
# methods defined in a scope such as `Kernel#global_variables` and
# `Kernel#local_variables` return an array of strings instead of an array of
# symbols like in later Ruby versions.

--- a/class.c
+++ b/class.c
@@ -1655,6 +1655,17 @@ class_instance_method_list(int argc, const VALUE *argv, VALUE mod, int obj, int
     st_foreach(me_arg.list, func, ary);
     st_free_table(me_arg.list);
 
+    /* If syntax transform is active and targeting Ruby <= 1.8, convert the output of all methods that list the names of instance methods in a class from an array of symbols to an array of strings */
+    if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+        long length = RARRAY_LEN(ary);
+        long i;
+        for (i = 0; i < length; ++i) {
+            VALUE value = rb_ary_entry(ary, i);
+            value = rb_sym2str(value);
+            rb_ary_store(ary, i, value);
+        }
+    }
+
     return ary;
 }
 
@@ -1892,6 +1903,17 @@ rb_obj_singleton_methods(int argc, const VALUE *argv, VALUE obj)
     st_foreach(me_arg.list, ins_methods_i, ary);
     st_free_table(me_arg.list);
 
+    /* If syntax transform is active and targeting Ruby <= 1.8, convert the output of `Object#singleton_methods` from an array of symbols to an array of strings */
+    if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+        long length = RARRAY_LEN(ary);
+        long i;
+        for (i = 0; i < length; ++i) {
+            VALUE value = rb_ary_entry(ary, i);
+            value = rb_sym2str(value);
+            rb_ary_store(ary, i, value);
+        }
+    }
+
     return ary;
 }
 
--- a/eval.c
+++ b/eval.c
@@ -409,7 +409,22 @@ rb_mod_s_constants(int argc, VALUE *argv, VALUE mod)
     if (cbase) {
 	data = rb_mod_const_of(cbase, data);
     }
-    return rb_const_list(data);
+    {
+        VALUE ary = rb_const_list(data);
+
+        /* If syntax transform is active and targeting Ruby <= 1.8, convert the output of `Module::constants` from an array of symbols to an array of strings */
+        if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+            long length = RARRAY_LEN(ary);
+            long i;
+            for (i = 0; i < length; ++i) {
+                VALUE value = rb_ary_entry(ary, i);
+                value = rb_sym2str(value);
+                rb_ary_store(ary, i, value);
+            }
+        }
+
+        return ary;
+    }
 }
 
 /*!
@@ -1868,7 +1883,20 @@ f_current_dirname(VALUE _)
 static VALUE
 f_global_variables(VALUE _)
 {
-    return rb_f_global_variables();
+    VALUE ary = rb_f_global_variables();
+
+    /* If syntax transform is active and targeting Ruby <= 1.8, convert the output of `Kernel#global_variables` from an array of symbols to an array of strings */
+    if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+        long length = RARRAY_LEN(ary);
+        long i;
+        for (i = 0; i < length; ++i) {
+            VALUE value = rb_ary_entry(ary, i);
+            value = rb_sym2str(value);
+            rb_ary_store(ary, i, value);
+        }
+    }
+
+    return ary;
 }
 
 /*
--- a/variable.c
+++ b/variable.c
@@ -1889,6 +1889,18 @@ rb_obj_instance_variables(VALUE obj)
 
     ary = rb_ary_new();
     rb_ivar_foreach(obj, ivar_i, ary);
+
+    /* If syntax transform is active and targeting Ruby <= 1.8, convert the output of `Object#instance_variables` from an array of symbols to an array of strings */
+    if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+        long length = RARRAY_LEN(ary);
+        long i;
+        for (i = 0; i < length; ++i) {
+            VALUE value = rb_ary_entry(ary, i);
+            value = rb_sym2str(value);
+            rb_ary_store(ary, i, value);
+        }
+    }
+
     return ary;
 }
 
@@ -2981,15 +2993,29 @@ VALUE
 rb_mod_constants(int argc, const VALUE *argv, VALUE mod)
 {
     bool inherit = true;
+    VALUE ary;
 
     if (rb_check_arity(argc, 0, 1)) inherit = RTEST(argv[0]);
 
     if (inherit) {
-	return rb_const_list(rb_mod_const_of(mod, 0));
+        ary = rb_const_list(rb_mod_const_of(mod, 0));
     }
     else {
-	return rb_local_constants(mod);
+        ary = rb_local_constants(mod);
+    }
+
+    /* If syntax transform is active and targeting Ruby <= 1.8, convert the output of `Module#constants` from an array of symbols to an array of strings */
+    if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+        long length = RARRAY_LEN(ary);
+        long i;
+        for (i = 0; i < length; ++i) {
+            VALUE value = rb_ary_entry(ary, i);
+            value = rb_sym2str(value);
+            rb_ary_store(ary, i, value);
+        }
     }
+
+    return ary;
 }
 
 static int
@@ -3696,7 +3722,22 @@ rb_mod_class_variables(int argc, const VALUE *argv, VALUE mod)
     else {
 	tbl = mod_cvar_at(mod, 0);
     }
-    return cvar_list(tbl);
+    {
+        VALUE ary = cvar_list(tbl);
+
+        /* If syntax transform is active and targeting Ruby <= 1.8, convert the output of `Object#class_variables` from an array of symbols to an array of strings */
+        if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+            long length = RARRAY_LEN(ary);
+            long i;
+            for (i = 0; i < length; ++i) {
+                VALUE value = rb_ary_entry(ary, i);
+                value = rb_sym2str(value);
+                rb_ary_store(ary, i, value);
+            }
+        }
+
+        return ary;
+    }
 }
 
 /*
--- a/vm_eval.c
+++ b/vm_eval.c
@@ -2521,7 +2521,22 @@ rb_f_local_variables(VALUE _)
 	    break;
 	}
     }
-    return local_var_list_finish(&vars);
+    {
+        VALUE ary = local_var_list_finish(&vars);
+
+        /* If syntax transform is active and targeting Ruby <= 1.8, convert the output of `Kernel#local_variables` from an array of symbols to an array of strings */
+        if (mkxp_ec_is_syntax_transform_active(1, 8, -1)) {
+            long length = RARRAY_LEN(ary);
+            long i;
+            for (i = 0; i < length; ++i) {
+                VALUE value = rb_ary_entry(ary, i);
+                value = rb_sym2str(value);
+                rb_ary_store(ary, i, value);
+            }
+        }
+
+        return ary;
+    }
 }
 
 /*
